File: questions/linkedList/reverse-linked-list.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function reverseList(head: ListNode | null): ListNode | null {
  if (!head) return head;

  // 1. set nodes
  let tailNode: ListNode | null = null;
  let headNode: ListNode | null = head;

  while (headNode) {
    // 2. preserve node
    const nextNode = headNode.next;
    // 3. set new pointer
    headNode.next = tailNode;
    // 4. move tail and head to the next
    tailNode = headNode;
    headNode = nextNode;
  }

  return tailNode;
}

File: questions/linkedList/reverse-linked-list.py
# Definition for singly-linked list.
from typing import Optional

class ListNode:
     def __init__(self, val=0, next=None):
         self.val = val
         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        tailNode, headNode = None, head

        while headNode:
            nextNode = headNode.next
            headNode.next = tailNode
            tailNode = headNode
            headNode = nextNode
        
        return tailNode